stdpopsim/engines.py

```diff
@@ -0,0 +1,122 @@
+import logging
+
+import attr
+import msprime
+import stdpopsim
+
+logger = logging.getLogger(__name__)
+
+_registered_engines = {}
+
+
+def register_engine(engine):
+    """
+    Registers the specified simulation engine.
+    """
+    logger.debug(f"Registering simulation engine '{engine.id}'")
+    _registered_engines[engine.id] = engine
+
+
+def get_engine(id):
+    """
+    Returns the simulation engine with the specified id.
+    """
+    if id not in _registered_engines:
+        raise ValueError(f"Simulation engine '{id}' not registered")
+    return _registered_engines[id]
+
+
+def all_engines():
+    """
+    Returns an iterator over all registered simulation engines.
+    """
+    for sim_engine in _registered_engines.values():
+        yield sim_engine
+
+
+@attr.s(frozen=True)
+class Engine(object):
+    """
+    Abstract class representing a simulation engine.
+
+    To implement a new simulation engine, one should inherit from this
+    class. At a minimum, the `id`, `name` and `citations` attributes must
+    be set, and the `simulate()` and `get_version()` methods must be
+    implemented. See msprime example in `engines.py`.
+
+    :ivar id The unique identifier for the simulation engine.
+    :vartype id: str
+    :ivar name: The informal name for this engine as it would be used in
+        written text such as the CLI or error messages.
+    :vartype name: str
+    :ivar citations: A list of citations for the simulation engine.
+    :vartype citations: list of :class:`.Citation`
+    """
+
+    def simulate(self, model=None, contig=None, samples=None, **kwargs):
+        """
+        Function to execute the actual simulation.
+
+        :var model
```

Comment:
Sorry, didn't spot this earlier --- this should be ``:param Model model: The model to simulate``. I think the ``var`` sphinx types are for instance variables. See below also.